pkg/datapath/linux/probes/probes.go

```diff
@@ -487,3 +434,171 @@ func HaveV3ISA() error {
+		if HaveProgramHelper(hp.ProgramType, hp.Helper) == nil {
+			probes.Helpers[hp] = true
+		}
+	}
+
+	if HaveLargeInstructionLimit() == nil {
+		probes.HaveLargeInsnLimit = true
+	}
+
+	return &probes
+}
+
+func writeCommonHeader(writer io.Writer, probes *HeaderProbes) error {
+	features := map[string]bool{
+		"HAVE_LRU_HASH_MAP_TYPE": probes.Maps[ebpf.LRUHash],
+		"HAVE_LPM_TRIE_MAP_TYPE": probes.Maps[ebpf.LPMTrie],
+		"HAVE_NETNS_COOKIE": probes.Helpers[HelperProbe{ebpf.CGroupSock, asm.FnGetNetnsCookie}] &&
+			probes.Helpers[HelperProbe{ebpf.CGroupSockAddr, asm.FnGetNetnsCookie}],
+		"HAVE_SOCKET_COOKIE": probes.Helpers[HelperProbe{ebpf.CGroupSockAddr, asm.FnGetSocketCookie}],
+		"HAVE_JIFFIES": probes.Helpers[HelperProbe{ebpf.CGroupSock, asm.FnJiffies64}] &&
+			probes.Helpers[HelperProbe{ebpf.CGroupSockAddr, asm.FnJiffies64}] &&
+			probes.Helpers[HelperProbe{ebpf.SchedCLS, asm.FnJiffies64}] &&
+			probes.Helpers[HelperProbe{ebpf.XDP, asm.FnJiffies64}],
+		"HAVE_SOCKET_LOOKUP": probes.Helpers[HelperProbe{ebpf.CGroupSockAddr, asm.FnSkLookupTcp}] &&
+			probes.Helpers[HelperProbe{ebpf.CGroupSockAddr, asm.FnSkLookupUdp}],
+		/* Before upstream commit d71962f3e627 (4.18), map helpers were not
+		 * allowed to access map values directly. So for those older kernels,
+		 * we need to copy the data to the stack first.
+		 * We don't have a probe for that, but the bpf_fib_lookup helper was
+		 * introduced in the same release. */
+		"HAVE_DIRECT_ACCESS_TO_MAP_VALUES": probes.Helpers[HelperProbe{ebpf.SchedCLS, asm.FnFibLookup}],
+		"HAVE_LARGE_INSN_LIMIT":            probes.HaveLargeInsnLimit,
+	}
+
+	return writeFeatureHeader(writer, features, true)
+}
+
+func writeSkbHeader(writer io.Writer, probes *HeaderProbes) error {
+	featuresSkb := map[string]bool{
+		"HAVE_CHANGE_TAIL": probes.Helpers[HelperProbe{ebpf.SchedCLS, asm.FnSkbChangeTail}],
+		"HAVE_FIB_LOOKUP":  probes.Helpers[HelperProbe{ebpf.SchedCLS, asm.FnFibLookup}],
+		"HAVE_CSUM_LEVEL":  probes.Helpers[HelperProbe{ebpf.SchedCLS, asm.FnCsumLevel}],
+	}
+
+	return writeFeatureHeader(writer, featuresSkb, false)
+}
+
+func writeXdpHeader(writer io.Writer, probes *HeaderProbes) error {
+	featuresXdp := map[string]bool{
+		"HAVE_FIB_LOOKUP": probes.Helpers[HelperProbe{ebpf.XDP, asm.FnFibLookup}],
+	}
+
+	return writeFeatureHeader(writer, featuresXdp, false)
+}
+
+func writeFeatureHeader(writer io.Writer, features map[string]bool, common bool) error {
+	const content = `
+/* SPDX-License-Identifier: (GPL-2.0-only OR BSD-2-Clause) */
+/* Copyright Authors of Cilium */
+
+/* THIS FILE WAS GENERATED DURING AGENT STARTUP. */
+
+#pragma once
+
+{{- if not .Common}}
+#include "features.h"
+{{- end}}
+
+{{- range $key, $value := .Features}}
+{{- if $value}}
+#define {{$key}} 1
+{{end}}
+{{- end}}
+`
+
+	input := struct {
+		Common   bool
+		Features map[string]bool
+	}{
+		Common:   common,
+		Features: features,
+	}
+
+	tpl, err := template.New("header").Parse(content)
```

Comment:
I'd make `tpl` a package var and call `.Parse()` from an `init()` at the top. `log.Fatal()` the error.